origin-js/src/resources/marketplace.js

```diff
@@ -244,52 +252,74 @@ export default class Marketplace {
-    //   the units purchased exceeds the units available at the time of the offer.
-    let unitsAvailable = listing.unitsTotal
+    // Step 2:
+    // This is unit listing specific. Filter out offers for which the units purchased exceeds 
+    // the units available or boost exceeds the expected boost at the time of the offer.
+    const isMultiUnit = listing.unitsTotal > 1 && listing.type === 'unit'
     const commission = listing.commission.amount
       ? BigNumber(listing.commission.amount)
       : null
-    const commissionPerUnit = listing.commissionPerUnit
+    const commissionPerUnit = isMultiUnit
       ? BigNumber(listing.commissionPerUnit.amount)
       : commission
-    const offers = []
-    allOffers.forEach(offer => {
-      const unitsSoldBeforeOffer = listing.unitsTotal - unitsAvailable
-
-      // Validate units purchased against units available.
-      // TODO: Validate units purchased vs. units available at offer creation
-      // time. This will handle the case where an edit of a listing invalidates
-      // a previously valid offer.
-      if (offer.unitsPurchased > unitsAvailable) {
-        return
-      }
-      if (offer.status !== 'withdrawn') {
-        unitsAvailable -= offer.unitsPurchased
-      }
+    let commissionAvailable = new BigNumber(commission) // create new instance
+    let unitsAvailable = listing.unitsTotal
+
+    const offers = allOffers.filter(offer => {
+      const offerCommission = offer.commission && BigNumber(offer.commission.amount)
 
-      // Validate that the offer commission is what we expect. If the amount
-      // of commission for the listing isn't sufficient for this offer, we
-      // require that the offer have whatever commission is available to it.
-      if (commissionPerUnit !== null && commissionPerUnit.isGreaterThan(0)) {
-        const commissionUsed = BigNumber(commissionPerUnit).times(unitsSoldBeforeOffer)
-        const remainingCommission = BigNumber.max(
-          commission.minus(commissionUsed),
+      // required for multi-unit commission and unitsAvailable calculations
+      const onValidOffer = () => {
+        if (offer.status === 'withdrawn')
+          return
+
+        unitsAvailable -= offer.unitsPurchased
+        commissionAvailable = BigNumber.max(
+          commissionAvailable.minus(offerCommission),
           BigNumber(0)
         )
-        const expectedCommission = BigNumber.min(
-          remainingCommission,
-          commissionPerUnit.times(offer.unitsPurchased)
-        )
-        const offerCommission = offer.commission && BigNumber(offer.commission.amount)
-        if (!offerCommission || !offerCommission.isEqualTo(expectedCommission)) {
-          return
+      }
+      /* Validate units purchased against units available.
+       *
+       * TODO: Validate units purchased vs. units available at offer creation
+       * time. This will handle the case where an edit of a listing invalidates
```

Comment:
I don't understand how validating units during offer creation handles this case. We do handle this case in `updateListing` though. See: https://github.com/OriginProtocol/origin/blob/master/origin-js/src/resources/marketplace.js#L395

Reply:
Yeah this comment is off. Let me change it to: 
```
Validate units purchased against units available.

This check is required because of an edge case, where 2 users can make an offer at the same time for the remaining amount of units in a listing. Both offers will get mined, since they are both valid at the time transaction to the blockchain is issued. After they are mined, only the first one is valid.
```